### PR TITLE
Use Deleter in more places, take 3

### PR DESCRIPTION
**Pull request by lptr**
- [x] cache directory cleanup
- [x] Groovy script compilation output

----
/cc @gradle/build-cache.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java [hunk 1]
```diff
@@ -75,26 +78,35 @@
...
 
     @Override
     public void compileToDir(ScriptSource source, ClassLoader classLoader, File classesDir, File metadataDir, CompileOperation<?> extractingTransformer,
                              Class<? extends Script> scriptBaseClass, Action<? super ClassNode> verifier) {
         Timer clock = Time.startTimer();
-        GFileUtils.deleteDirectory(classesDir);
-        GFileUtils.mkdirs(classesDir);
+        try {
+            deleter.ensureEmptyDirectory(classesDir, true);
+        } catch (IOException ioex) {
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java]: Can we make the methods of `Deleter` not throw an IOException? Throwing an unchecked exception would make the usage easier.

File: subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java [hunk 2]
```diff
@@ -78,6 +78,7 @@ static WorkResult execute(ServiceRegistry serviceRegistry, final Iterable<File>
             }
 
             if (spec.getScalaCompileOptions().isForce()) {
+                // TODO This should use Deleter
```
Comment (wolfs) [inline subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java]: Will this be fixed in a follow up?

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java]: I looked at that and in several places it does make sense to catch the exception and add more context. Let's keep it this way for now.

File: subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java (hunk 2 above)
Comment (lptr) [reply subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java]: I added https://github.com/gradle/gradle/issues/10277. This is a bit more complicated as `Deleter` would need to be exposed to workers etc.

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java]: Nobody stops you from catching the exception to add more context even if we throw only an unchecked exception.

**Pull request merged**